### DISCUSSION TITLE
License questions

### DISCUSSION THREAD

**Discussion by Aspie96** [4ian/GDevelop#2364] [Enhancement ideas (and their technical discussions)]
Hi.

I have a few questions and suggestions relative to the licensing of GDevelop.

## LGPL license
Most of GDevelop computer code is under the MIT license and other permissive free licenses.

The only part which I could find under a copyleft license is glui in Box2D, under LGPL.

However, the current version of Box2D does not contain glui: https://github.com/erincatto/box2d.

In addition, the current version of glui is under zlib, instead of LGPL: that is a permissive license instead of a copyleft one: https://github.com/libglui/glui.

I don't know the codebase of GDevelop, so I do not know how easy or hard this would be. However, I suggest either switching to the latest version of Box2D or switching to the latest version of glui.

Either way, this avoids having part of this software as copyleft, which is good, given that the intention is clearly that of it being permissively licensed and the old version of Box2D and glui seems to be the only obstacle to that.

## Liberation font
Besides code, one non-permissive copyleft component is the Liberation font, under SIL OFL, which is a free copyleft license.
I suggest mentioning this component in the main License file, just to make sure the user is aware there is at least one copyleft component.
I couldn't find where the font is actually stored, to be honest.

Besides, files in /newIDE/app/resources/examples/pairs/images are under CC BY-SA 4.0, also copyleft. I suggest mentioning it in the root license file as well.

## CC 4.0 icon
In addition, I noticed this project uses the following icon: https://iconify.design/icon-sets/bx/bxs-network-chart.html, added in https://github.com/4ian/GDevelop/pull/1842#issuecomment-652449557.
This icon is "CC" licensed, which means absolutely nothing.
There are many Creative Commons licenses, most of which are not free. "Creative Commons" is just an organization which writes licenses. Confusion among them is a common mistake.

The page does not mention which license the icon is, but I found it is likely the Creative Commons Attribution 4.0 International Public License: I explained why in an issue there: https://github.com/atisawd/boxicons/issues/111.

So far, I see no attribution for such icon anywhere in GDevelop, and it's required by all CC licenses. While they do write "Attribution is not required but is appreciated". It's not clear to me whether they refer to just the font or both the font and the icons, but I think it's likely they refer to both.

Still, this license should be mentioned somewhere because the CC-BY 4.0 licenses do have some interesting restrictions beyond the attribution clause.

## Cordova
I notice some parts of the repo mention Cordova. Are parts of Cordova contained in the repo? If so, the Apache license should be mentioned.

## License change
Finally, I noticed parts of this project were under either GPL or LGPL, but are now under permissive licenses. May I ask how the license change happened? I was unable to find mentions of it.

**Comment by 4ian** [comment]
Hi, thanks for checking this :)
See my answers below, feel free to open a PR if something can be fixed quickly this way.

> However, the current version of Box2D does not contain glui: https://github.com/erincatto/box2d.

In GDevelop 5, we're using the Box2d.js port (https://github.com/kripken/box2d.js/) which is zlib licensed, so there should not be any usage of glui.
We used to have Box2d, the C++ version, [in GDevelop 4](https://github.com/4ian/GDevelop/tree/master/Extensions/PhysicsBehavior/Box2D/Box2D), but glui was never used. 

It's true that because it's still in the source tree, it might be confusing - at some point we'll have to clean up all the unused code in GDevelop 5.
In all cases, we should be good as we're never linking to/using/redistributing glui.

> Liberation font

This was used in GDevelop 4, not used anymore even if still in the source tree. I see the license is written in the file itself, but might be worth removing the file entirely then.

> CC 4.0 icon

@arthuro555 would you be able to quickly look at this and see if we can add the attribution in a license file next to it?
(we can then add a link to this license file from the root license file)

> I notice some parts of the repo mention Cordova. Are parts of Cordova contained in the repo? If so, the Apache license should be mentioned.

Not in this repo, Cordova is an entirely separate project, not embedded in GDevelop.

> Besides, files in /newIDE/app/resources/examples/pairs/images are under CC BY-SA 4.0, also copyleft. I suggest mentioning it in the root license file as well.

Good catch, @Wend1go worked on this example I think, could you check if we have a license file in the folder of this example? Let's add the CC BY-SA 4.0 there if it's missing :)

> Finally, I noticed parts of this project were under either GPL or LGPL, but are now under permissive licenses. May I ask how the license change happened? I was unable to find mentions of it.

The old editor of GDevelop 4 was GPL - all the code was removed (folder `IDE` at root, which is now gone since a few years) when transitioning to GD5 (folder `newIDE` at root, which is MIT licensed).
GDevelop 5 editor was remade from scratch, and no part of the old editor were reused (notably because the tech stacks are entirely different).

It's possible that I started the game engine (`GDCpp`) and the core (`GDCore`) with LGPL as a license (I would have to check). At the time having no contributors (or maybe just @victorlevasseur), so I moved to MIT without issues :)

**Comment by arthuro555** [comment]
> would you be able to quickly look at this and see if we can add the attribution in a license file next to it?

I assumed by default that it was CC0 as it wasn't written with the other suffices (BY, SA, NC), but indeed it isn't clear. The link to the full text license in their GitHub is broken, so we would have to ask the author directly, what Aspie96 already did. The most confusing thing is that in their package.json they have CC-BY-4.0 according to Aspie96 but their website says that there is no attribution needed.

**Comment by Aspie96** [comment]
Thank you for taking the time answering this!

> It's true that because it's still in the source tree, it might be confusing - at some point we'll have to clean up all the unused code in GDevelop 5.
> In all cases, we should be good as we're never linking to/using/redistributing glui.

In this case, I suggest removing it, since it provides no actual benefit but may be confusing for what legal requirements apply, if one looks for license files.

> This was used in GDevelop 4, not used anymore even if still in the source tree. I see the license is written in the file itself, but might be worth removing the file entirely then.

I'd say the same applies: it's confusing, unless somebody analyzes the code.

> Not in this repo, Cordova is an entirely separate project, not embedded in GDevelop.

Thanks for clarifying.

> Good catch, @Wend1go worked on this example I think, could you check if we have a license file in the folder of this example? Let's add the CC BY-SA 4.0 there if it's missing :)

CC licenses do not ask that a full copy of the license is included, so that's up to you.

Given CC BY-SA 4.0 is, at this point, the most restrictive license in the project, though, I think it should be mentioned in the main license file of GDevelop, just so the user is aware there is a copyleft component :-)

> I assumed by default that it was CC0

I checked and it is almost defenitely not CC0, but CC BY-4.0.

You can see it in the package.json. Plus, it's what their website links to when saying "CC 4.0". Also consider that "Creative Commons License" is used by many to confusingly refer to CC BY, for some reason. Plus, CC0 doesn't have a 4.0 version.

> but their website says that there is no attribution needed.

This is confusing to me as well, but it might just be an additional exception: if so, other conditions still apply. Including attribution is still a good idea because it's a reference that can be useful, even if it's not required.

**Comment by Wend1go** [comment]
> CC licenses do not ask that a full copy of the license is included, so that's up to you.
> 
> Given CC BY-SA 4.0 is, at this point, the most restrictive license in the project, though, I think it should be mentioned in the main license file of GDevelop, just so the user is aware there is a copyleft component :-)

I don't know if this is necessary since the license file gets copied into the project folder when the example is opened in the IDE. So it always stays with the project files.
I made the graphics myself for a game that I already published so I just took the same license that I used for the game assets.
I'm perfectly fine with the way it is handled right now, so no worries. 😉

**Comment by Aspie96** [comment]
> I don't know if this is necessary since the license file gets copied into the project folder when the example is opened in the IDE. So it always stays with the project files.

EDIT:

While I disagreed with this at first, I agree now, you are right about this.
I still think, though, it should be mentioned that examples contain assets from other sources, so the user is aware they have to check this
